astropy/units/quantity.py

```diff
@@ -286,6 +315,99 @@ class Quantity(np.ndarray):
 
     __array_priority__ = 10000
 
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+
+        # Register the TypeVar for cached reuse.
+        # See __class_getitem__ for usage.
+        _quantitytype_registry[repr(cls)] = T.TypeVar(repr(cls), bound=cls)
+
+    def __class_getitem__(cls, unit_annotation):
+        """Quantity Type Hints.
+
+        .. |Unit| replace:: :class:`~astropy.units.core.UnitBase`
```

Comment:
I do have a solution for that. In ``docs/conf.py`` the "rst_epilog" can hold all substitutions. These substitutions may be used in the docs AND in any docstring. So I use |Unit| and |Quantity| and other very obvious substitutions quite liberally.

Reply:
I like that idea *a lot* -- but let's do it outside of this PR. This would affect lots and lots of docstrings and we have to be sure people like the idea enough. E.g., numpy uses "array-like" for many of its docstrings and possibly it is nicer to do something similar (in many cases, a `Column` with a `unit` will just work where a `Quantity` would to, or a string for a unit.

Reply:
https://github.com/astropy/astropy/issues/11063

Reply:
Well that was a whole other can of worms!